util/rate_limiter.cc

```diff
@@ -256,15 +274,55 @@ int64_t GenericRateLimiter::CalculateRefillBytesPerPeriod(
   }
 }
 
+Status GenericRateLimiter::Tune() {
+  const int kLowWatermarkPct = 50;
+  const int kHighWatermarkPct = 90;
+  const int kAdjustFactorPct = 5;
+  // computed rate limit will be in
+  // `[max_bytes_per_sec_ / kAllowedRangeFactor, max_bytes_per_sec_]`.
+  const int kAllowedRangeFactor = 20;
+
+  std::chrono::microseconds prev_tuned_time = tuned_time_;
+  tuned_time_ = std::chrono::microseconds(NowMicrosMonotonic(env_));
+
+  int64_t elapsed_intervals = (tuned_time_ - prev_tuned_time +
+                               std::chrono::microseconds(refill_period_us_) -
+                               std::chrono::microseconds(1)) /
+                              std::chrono::microseconds(refill_period_us_);
+  int64_t drained_pct =
+      (num_drains_ - prev_num_drains_) * 100 / elapsed_intervals;
+  int64_t prev_bytes_per_sec = GetBytesPerSecond();
+  int64_t new_bytes_per_sec;
+  if (drained_pct == 0) {
+    new_bytes_per_sec = max_bytes_per_sec_ / kAllowedRangeFactor;
+  } else if (drained_pct < kLowWatermarkPct) {
+    new_bytes_per_sec =
+        std::max(max_bytes_per_sec_ / kAllowedRangeFactor,
+                 prev_bytes_per_sec * 100 / (100 + kAdjustFactorPct));
```

Comment:
In this function, needs to care about overflow in general. For example, `prev_bytes_per_sec * 100` can technically overflow. It is a possible that in somewhere else, the value of GetBytesPerSecond() has been sanitized so that it is not close to overflow here, but it's not super clear when I read this function. Can user set a very high rate limiter threshold, like max_int and it gets weird results here? It ever caused problem to rate limiter before. Some users think if they don't need to limit the rate, just set it to max_int. So I suggest it is explicitly handled in the code here.

Reply:
Sure, I will check whether the multiplication would overflow, and if so, set the new rate limit to max int. I vaguely remember you wrote similar code before.